### ISSUE TITLE
Manage feature dependencies

### ISSUE THREAD

**Issue by triplem** [ktorio/ktor#222]
This is possible but couples the Features top much. There should be a way top allow accessing vals from one feature in another one without top tight coupling.

A possible Use case are eg. the metrics feature and the sessions feature to allow metrics on the number of sessions.

**Comment by orangy** [comment]
You can get a Feature instance from application. There are some quirks if you install features deeper in the Routing, but for top level features it's okay.

**Comment by triplem** [comment]
Let me rephrase the use case somehow. Wouldn't it be nice, if a feature is using other features if those are enabled out if the box? Eg a session metric could be provided by ktor automatically if both features are enabled. Otherwise some kind of meta features are possibly programmed again and again...

What do you think?

**Comment by orangy** [comment]
Feature dependency is indeed much needed thing, but we don't have a design yet.

**Comment by triplem** [comment]
This should take into account that two features are implementing pretty much the same thing (eg gson and Jackson). Some dependency issues are coming up then.

**Comment by thecoshman** [comment]
As a stop gap measure, we could do with a clear convention for features stating what other features they depend on.

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-820">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
